### PR TITLE
feat: coverage in mainline CI

### PR DESCRIPTION
**Pull request by smartcontracts**
Adds converage to mainline CI.

### CODE REVIEW

File: codecov.yml
Comment (mds1) [inline codecov.yml]: Think you meant 100 here? Though maybe 95 is better, some branches can be really hard to get coverage for when they are never intended to be hit (e.g. invariants). Or, we can start at 100, and loosen if it becomes a blocker
```suggestion
        threshold: 100%
```

Comment (tynes) [reply codecov.yml]: I definitely lean in favor of not 100% because universally not all loc will need to be tested and then it creates this artificial headache blocking merge

Comment (smartcontracts) [reply codecov.yml]: Let's start at 100% and loosen if it's a blocker. Codecov is backwards and this threshold represents how much coverage is allowed to drop, so 0% is correct here.

Comment (smartcontracts) [reply codecov.yml]: Can we agree on loosening this to 95%?

**Review by mds1 — APPROVED** [review]

File: codecov.yml
Comment (mds1) [reply codecov.yml]: Approved the PR, pending this change

Comment (tynes) [reply codecov.yml]: You own this code so you can do whatever you want here, I do think 100% is not a good idea and would prefer 95%

Comment (tynes) [reply codecov.yml]: https://about.codecov.io/blog/the-case-against-100-code-coverage/

>Although there is no perfect code coverage value, we can still make a stab at such a value. 2 years ago, Google posted a [blog post](https://testing.googleblog.com/2020/08/code-coverage-best-practices.html) on their best practices:
“We offer the general guidelines of 60% as “acceptable”, 75% as “commendable” and 90% as “exemplary.””
At Codecov, we see that a significant portion of our open source repositories centers around 80% coverage.

Comment (smartcontracts) [reply codecov.yml]: Ok let's start with 95% and see what happens. We can adjust if necessary, will be flexible with this.

Comment (smartcontracts) [reply codecov.yml]: @tynes can you approve in that case?

**Review by tynes — APPROVED** [review]

**Pull request merged**
